master/docs/developer/py3-compat.rst

```diff
@@ -0,0 +1,74 @@
+Python3 compatibility
+=====================
+
+
+A good place to start looking for advice to ensure that any code is compatible with both python-3.x and python2.7/2.6
+is too look at the python-future `cheat sheet <http://python-future.org/compatible_idioms.html>`_ .
+Buildbot uses the python-future library to ensure compatability with both python2.7/2.6 and python3.x.
+
+
+Dictionaries
+------------
+In python3, ``dict.iteritems`` is no longer available. While ``dict.items()`` does exist, it can be memory
+itensive in python2. For this reason, please use the python.future function ``iteritems()``
+
+Example::
+
+    d = {"cheese": 4, "bread": 5, "milk": 1}
+    for item, num in d.iteritems():
+        print("We have {} {}".format(num, item))
+
+should be written as::
+
+    from future.utils import iteritems
+    d = {"cheese": 4, "bread": 5, "milk": 1}
+    for item, num in iteritems(d):
+        print("We have {} {}".format(num, item))
+
+This also applies to the similar methods ``dict.itervalues()`` and ``dict.values()``, which have an equivalent ``itervalues()``
+
+If a list is required, please use ``list(iteritems(dict))``. This is for compatibility with the six library
+
+For iterating over dictionary keys, please use ``for key in dict:``.
+For example::
+
+    d = {"cheese": 4, "bread": 5, "milk": 1}
+    for item in d:
+        print("We have {}".format(item))
+
+Similarly when you want a list of keys::
+
+    keys = list(d)
+
+Strings
+-------
+In python2, there is only one type of string. It can be both unicode and bytestring. In python3, this is
+no longer the case. For this reasons all string must be marked with either ``u''`` or ``b''`` to signify if
+the string is a unicode string or a bytestring respectively
+
+Example::
+
+    u'this is a unicode string, a string for humans to read'
+    b'This is a bytestring, a string for computers to read'
+
+Print statements
+----------------
+Print statements are gone in python3. Please import ``from __future__ import print_function`` at the very top
+of the module to enable use of python3 style print functions
+
+Exceptions
+----------
+All exceptions should be written with the ``as`` statement. Before::
```

Comment:
I think it's a good idea to keep it.

As for "should" vs "must", I'd say that _new_ code it can be 'must'.

As for enforcing, I wonder if there's a tool to perform the check.